lib/net/http/header.rb

```diff
@@ -806,56 +806,107 @@ def set_form_data(params, sep = '&')
+  #
+  # The form data given in +params+ consists of zero or more fields;
+  # each field is:
+  #
+  # - A scalar value.
+  # - A name/value pair.
+  # - An IO stream opened for reading.
+  #
+  # Argument +params+ should be an
+  # {Enumerable}[https://docs.ruby-lang.org/en/master/Enumerable.html#module-Enumerable-label-Enumerable+in+Ruby+Classes]
+  # (method <tt>params.map</tt> will be called),
+  # and is often an array or hash.
+  #
+  # First, we set up a request:
+  #
+  #   _uri = uri.dup
+  #   _uri.path ='/posts'
+  #   req = Net::HTTP::Post.new(_uri)
+  #
+  # <b>Argument +params+ As an Array</b>
+  #
+  # When +params+ is an array,
+  # each of its elements is a subarray that defines a field;
+  # the subarray may contain:
+  #
+  # - One string:
+  #
+  #     req.set_form([['foo'], ['bar'], ['baz']])
+  #     # String to be included in request body: 'foo&bar&baz'
+  #
+  # - Two strings:
+  #
+  #     req.set_form([%w[foo 0], %w[bar 1], %w[baz 2]])
+  #     # String to be include in request body: 'foo=0&bar=1&baz=2'
+  #
+  # - A string (ignored) and an IO stream opened for reading:
+  #
+  #     File.write('t.tmp', 'Ruby is cool.')
+  #     file = File.open('t.tmp')
+  #     req.set_form([['not used', file]])
+  #
+  # - A string (ignored), an IO stream opened for reading,
+  #   and an options hash, which may contain these entries:
   #
   #   - +:filename+: The name of the file to use.
   #   - +:content_type+: The content type of the uploaded file.
   #
-  # Each item is a file field or a normal field.
-  # If +value+ is a File object or the +opt+ hash has a :filename key,
-  # the item is treated as a file field.
+  #   Example:
+  #
+  #     req.set_form([['not used', file, {filename: "other-filename.foo"}]]
+  #
+  # The various forms may be mixed:
+  #
+  #   req.set_form(['foo', %w[bar 1], ['not used', file]])
+  #
+  # <b>Argument +params+ As a Hash</b>
+  #
+  # When +params+ is a hash,
+  # each of its entries is a name/value pair that defines a field:
+  #
+  # - The name is a string.
+  # - The value may be:
+  #
+  #   - +nil+.
+  #   - Another string.
+  #   - An IO stream opened for reading.
+  #
+  # Examples:
+  #
+  #   # Nil-valued fields.
+  #   req.set_form({foo: nil, bar: nil, baz: nil})
+  #   # String to be included in request body: 'foo&bar&baz'
+  #
+  #   # String-valued fields.
+  #   req.set_form({foo: 0, bar: 1, baz: 2})
+  #   # String to be include in request body: 'foo=0&bar=1&baz=2'
   #
-  # If Transfer-Encoding is set as chunked, this sends the request using
-  # chunked encoding. Because chunked encoding is HTTP/1.1 feature,
-  # you should confirm that the server supports HTTP/1.1 before using
-  # chunked encoding.
+  #   # IO-valued field.
+  #   File.write('t.tmp', 'Ruby is cool.')
+  #   file = File.open('t.tmp')
+  #   req.set_form({not_used: file})
   #
-  # Example:
+  #   # Mixture of fields.
+  #   req.set_form({foo: nil, bar: 1, not_used: file})
```

Comment:
These examples use a symbol as the key of the hash, this doesn't work for `multipart/form-data`.

```ruby
req.send(:encode_multipart_form_data, io, {foo: "bar"}, {})
# => undefined method `gsub' for :foo:Symbol (NoMethodError)
```

Reply:
Fixed.